tests/sound/sound.c

```diff
@@ -0,0 +1,278 @@
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* Sound request */
+enum {
+    INIT_SOUND,
+    SHUTDOWN_SOUND,
+    PLAY_MUSIC,
+    PLAY_SFX,
+    SET_MUSIC_VOLUME,
+    STOP_MUSIC,
+};
+
+typedef struct {
+    void *data;
+    int size;
+} musicinfo_t;
+
+typedef struct sfxinfo_struct {
+    void *data;
+    int size;
+} sfxinfo_t;
+
+#define MUSIC_MAX_SIZE 65536
+#define SFX_MAX_SIZE 32768
+
+sfxinfo_t *sfx;
+musicinfo_t *music;
+
+static char *sfx_src;
+static char *music_src;
+static int music_delay = 30;
+static int music_looping = 0; /* FIXME: parsing looping CLI argument */
+static int music_volume = 3;
+static int sfx_volume = 15;
+
+static bool has_music = false;
+static bool has_sfx = false;
+static bool has_incr_music_volume = false;
+static bool has_sfx_repeat = false;
+
+void play_music()
+{
+    int request_type = PLAY_MUSIC;
+
+    register int a0 asm("a0") = request_type;
+    register int a1 asm("a1") = (uintptr_t) music;
+    register int a2 asm("a2") = music_looping;
+    register int a3 asm("a3") = music_volume;
+    register int a7 asm("a7") = 0xBABE;
+
+    asm volatile("scall" : "+r"(a0) : "r"(a1), "r"(a2), "r"(a3), "r"(a7));
+}
+
+void stop_music()
+{
+    int request_type = STOP_MUSIC;
+
+    register int a0 asm("a0") = request_type;
+    register int a7 asm("a7") = 0xBABE;
+
+    asm volatile("scall" : "+r"(a0) : "r"(a7));
+}
+
+void play_sfx()
+{
+    int request_type = PLAY_SFX;
+
+    register int a0 asm("a0") = request_type;
+    register int a1 asm("a1") = (uintptr_t) sfx;
+    register int a2 asm("a2") = sfx_volume;
+    register int a7 asm("a7") = 0xBABE;
+
+    asm volatile("scall" : "+r"(a0) : "r"(a1), "r"(a2), "r"(a7));
+}
+
+void set_music_volume()
+{
+    int request_type = SET_MUSIC_VOLUME;
+
+    register int a0 asm("a0") = request_type;
+    register int a1 asm("a1") = music_volume;
+    register int a7 asm("a7") = 0xBABE;
+
+    asm volatile("scall" : "+r"(a0) : "r"(a1), "r"(a7));
+}
+
+/*
+ * when emulating the Doom, sfx is in lump
+ * so sfx_src should be a valid sfx lump
+ */
+static void load_sfx()
+{
+    sfx = malloc(sizeof(sfxinfo_t));
+    assert(sfx != NULL);
+    sfx->data = malloc(sizeof(uint8_t) * SFX_MAX_SIZE);
+    assert(sfx->data != NULL);
+    sfx->size = 5669; /* FIXME: hardcoded size */
+
+    FILE *_sfx = fopen(sfx_src, "rb");
```

Comment:
Avoid the naming scheme starting with underscore.

Here is what the C standard states (section 7.1.3):
* All identifiers that start with an underscore followed by an uppercase letter or another underscore are always reserved for any purpose.
* All identifiers that start with an underscore are always reserved to be used as identifiers with file scope in both the ordinary and tag name spaces.

(The section proceeds to enumerate specific identifiers and groups of identifiers that are reserved by certain standard headers.)

The implication is that, for example, the implementation (whether it's the compiler or a standard header) can utilize the name `__FOO` for any purpose it deems fit.

In light of this, it's best to abstain from defining any such identifiers in your own code (unless your code is part of a C implementation -- and if you are uncertain, it likely is not). There's no real necessity to define these identifiers anyway; there is an abundance of unreserved identifiers at your disposal.

You can employ an identifier like `_foo` as long as it is confined to a local definition (not at file scope). However, I personally find it much simpler to avoid the use of leading underscores entirely.